File: packages/node/src/Seguranca/Ataques/AtaqueColisaoIV.js
'use strict';

const ResultadoAtaque = require('../ResultadoAtaque.js');
const SkipAtaqueException = require('../SkipAtaqueException.js');
const CriptografiaAlvo = require('../CriptografiaAlvo.js');

/**
 * Se o mesmo IV aparece duas vezes com a mesma KEY, a cifra perde as
 * garantias de confidencialidade (keystream reusado = "two-time pad").
 * Gera muitos tokens do MESMO texto e verifica se os IVs nunca colidem.
 */
class AtaqueColisaoIV {
  constructor({ geracoes = 5000 } = {}) {
    this.geracoes = geracoes;
  }

  nome() {
    return 'Colisão de IV';
  }

  executar(alvo) {
    if (!(alvo instanceof CriptografiaAlvo)) {
      throw new SkipAtaqueException('Precisa de base64url_decode do alvo.');
    }

    const vistos = new Set();
    let colisoes = 0;

    for (let i = 0; i < this.geracoes; i++) {
      const token = alvo.encrypt('MESMO_TEXTO_SEMPRE');
      const decodificado = alvo.base64urlDecode(token.slice(alvo.prefixo().length));
      const iv = alvo.decompor(decodificado).iv;
      const chaveIv = iv.toString('hex');
      if (vistos.has(chaveIv)) {
        colisoes++;
      }
      vistos.add(chaveIv);
    }

    if (colisoes > 0) {
      return new ResultadoAtaque(
        this.nome(),
        true,
        'critica',
        `${colisoes} colisão(ões) de IV em ${this.geracoes} gerações`,
      );
    }

    return new ResultadoAtaque(
      this.nome(),
      false,
      'info',
      `0 colisões em ${this.geracoes} gerações`,
    );
  }
}

module.exports = AtaqueColisaoIV;

File: packages/python/src/Seguranca/Ataques/AtaqueColisaoIV.py
from ..CriptografiaAlvo import CriptografiaAlvo
from ..ResultadoAtaque import ResultadoAtaque
from ..SkipAtaqueException import SkipAtaqueException


class AtaqueColisaoIV:
    """
    Se o mesmo IV aparece duas vezes com a mesma KEY, a cifra perde as
    garantias de confidencialidade (keystream reusado = "two-time pad").
    Gera muitos tokens do MESMO texto e verifica se os IVs nunca colidem.
    """

    def __init__(self, geracoes: int = 5000):
        self.geracoes = geracoes

    def nome(self) -> str:
        return "Colisão de IV"

    def executar(self, alvo) -> ResultadoAtaque:
        if not isinstance(alvo, CriptografiaAlvo):
            raise SkipAtaqueException("Precisa de base64url_decode do alvo.")

        vistos = set()
        colisoes = 0

        for _ in range(self.geracoes):
            token = alvo.encrypt("MESMO_TEXTO_SEMPRE")
            decodificado = alvo.base64url_decode(token[len(alvo.prefixo()) :])
            iv = alvo.decompor(decodificado)["iv"]
            chave_iv = iv.hex()
            if chave_iv in vistos:
                colisoes += 1
            vistos.add(chave_iv)

        if colisoes > 0:
            return ResultadoAtaque(
                self.nome(),
                True,
                "critica",
                f"{colisoes} colisão(ões) de IV em {self.geracoes} gerações",
            )

        return ResultadoAtaque(
            self.nome(),
            False,
            "info",
            f"0 colisões em {self.geracoes} gerações",
        )
